### ISSUE TITLE
Help for Adversial example to apply it to a single image with a pre-trained network with Batch Normalization?

### ISSUE THREAD

**Issue by erogol** [apache/incubator-mxnet#895]
My real purpose is to load a pretrained network to the work-flow proposed in example/adversary and obtain a gradient image for a single image.

I load a pre-trained model by the following code and pass all its parameters to newly created model.

```
# Create new model graph
fc = alexnet_grad_img_model(21)
dev = mx.gpu(0)

# Load pretrained net
epoch = 25
model_prefix = "models/AlexNet"
tmp_model = mx.model.FeedForward.load(model_prefix, epoch, ctx=dev)

# Pass params of trained net to new model 
model_params = tmp_model.get_params()

data_shape = (batch_size, 3, 224, 224)
model_params['arg_params']['data'] = mx.nd.zeros(data_shape, ctx=dev)

arg_names = model_params['arg_params'].keys()
grad_arrays = [(name, mx.nd.zeros(param.shape, ctx=dev)) for name, param in model_params['arg_params'].items()]
grad_map = dict(grad_arrays)
model_params['args_grad'] = grad_map
model_params['grad_req'] = dict([(name,"write") for name in arg_names])

# Bind the new model
model = fc.bind(ctx=dev, args=model_params['arg_params'], args_grad= model_params['args_grad'], aux_states=model_params['aux_params'], grad_req= model_params['grad_req'])

data_grad = model_params["args_grad"]['data']
out_grad = mx.nd.zeros(model.outputs[0].shape, ctx=dev)
```

After this stage, even in everywhere I defined the context as mx.gpu(0), still model stays in cpu memory as the output of `print model.debug_str()` says. My first question is how can I keep the new model inside GPU?

The second problem is I use BN with AlexNet and when I reduce the batch_size to 1 due to BN (I believe), I got always skewed outputs and always same class prediction. However, if I increase the batch_size to 64 (which is my training batch size), prediction go smoothly. My second question is how can I handle this problem for single image prediction? Below is my stupid code I use for single image prediction with iterators created by mx.io.ImageRecordIter.

```
num_round = 4
train_acc = 0.
nbatch = 0
for i in range(num_round):
    train_loss = 0.
    train_acc = 0.
    nbatch = 0
    train_iter.reset()
    for data, label in train_iter:
        arg_map["data"][:] = data
        model.forward(is_train=True)
        theta = model.outputs[0].asnumpy()
        alpha = Softmax(theta)
        train_acc += CalAcc(alpha, label.asnumpy()) / batch_size
        train_loss += CalLoss(alpha, label.asnumpy()) / batch_size
        losGrad_theta = LogLossGrad(alpha, label.asnumpy())
        out_grad[:] = losGrad_theta
        model.backward([out_grad])
        # data_grad[:] = grad_map["data"]
        for name in arg_names:
            if name != "data":
                SGD(arg_map[name], grad_map[name])

        nbatch += 1
    #print(np.linalg.norm(data_grad.asnumpy(), 2))
    train_acc /= nbatch
    train_loss /= nbatch
    print("Train Accuracy: %.2f\t Train Loss: %.5f" % (train_acc, train_loss))
```

**Comment by piiswrong** [comment]
You'll need to modify the BN layer for that. Compute dataset-wise statistics for mean and std, and use them at test time instead of compute from batch everytime

**Comment by erogol** [comment]
You mean setting an parameter accordingly or coding this function to that layer. But how does indeption net predict on a single image in predict-with-pretrained-model.ipynb ?

**Comment by tqchen** [comment]
You want to turn ```is_train=False``` for test time prediction, and it should work for any batch size

**Comment by erogol** [comment]
But in that case I cannot get the gradients or as far as I give grad arrays to the networks they are computed by model.bacward()

**Comment by erogol** [comment]
Also could you give some comments for the way I try to copy a pre-trained model to a new one.

**Comment by tqchen** [comment]
All the model parameters are in the argument arrays, so should be straight-forward. Unfortunately since BN is a cross image operator so it is impossible to get gradient when you change batch size

**Comment by szha** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
